### PR TITLE
UCT/IB: Add MR to fence relaxed order puts

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/uct/ib/base/ib_md.c [hunk 1]
```diff
@@ -240,7 +240,8 @@ UCS_LIST_HEAD(uct_ib_md_ops_list);
 
 typedef struct uct_ib_verbs_mem {
     uct_ib_mem_t        super;
-    struct ibv_mr       *mr;
+    uct_ib_mr_t         mr;
+    uct_ib_mr_t         so_mr;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:244]: need better name

File: src/uct/ib/base/ib_md.c [hunk 2]
```diff
@@ -553,22 +554,41 @@ ucs_status_t uct_ib_dereg_mrs(struct ibv_mr **mrs, size_t mr_num)
...
+        s = md->ops->dereg_atomic_key(md, memh);
         memh->flags &= ~UCT_IB_MEM_FLAG_ATOMIC_MR;
+        if (s != UCS_OK) {
+            status = s;
+        }
+    }
+
+    if (memh->flags & UCT_IB_MEM_FLAG_RELAXED_ORDERING) {
+        s = uct_ib_memh_dereg_key(md, memh);
+        memh->flags &= ~UCT_IB_MEM_FLAG_RELAXED_ORDERING;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:580]: this looks weird, UCT_IB_MEM_FLAG_RELAXED_ORDERING means we need to call uct_ib_memh_dereg_key twice?

File: src/uct/ib/base/ib_md.c [hunk 3]
```diff
@@ -721,6 +741,16 @@ static ucs_status_t uct_ib_mem_reg_internal(uct_md_h uct_md, void *address,
         return status;
     }
 
+    if (md->relaxed_order) {
+        access_flags &= ~IBV_ACCESS_RELAXED_ORDERING;
+        status = uct_ib_md_reg_mr(md, address, length, access_flags, silent, memh);
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:746]: do not modify access_flags , pass access_flags  & ~IBV_ACCESS_RELAXED_ORDERING

File: src/uct/ib/base/ib_md.c [hunk 4]
```diff
@@ -721,6 +741,16 @@ static ucs_status_t uct_ib_mem_reg_internal(uct_md_h uct_md, void *address,
...
+        if (status == UCS_OK) {
+            memh->flags |= UCT_IB_MEM_FLAG_RELAXED_ORDERING;
+        } else {
+            goto err;
+        }
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:751]: if (status != OK)
 goto err

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 5]
```diff
@@ -15,53 +15,70 @@
...
+    struct mlx5dv_devx_obj     *dvmr;
     int                        mr_num;
     size_t                     length;
     struct ibv_mr              *mrs[];
 } uct_ib_mlx5_ksm_data_t;
 
+typedef union uct_ib_mlx5_mr {
+    uct_ib_mr_t                super;
+    uct_ib_mlx5_ksm_data_t     *ksm_data;
+} uct_ib_mlx5_mr_t;
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:27]: mlx5 data structure refactoring (removing union) should have been a separate PR

File: src/uct/ib/base/ib_md.h [hunk 6]
```diff
@@ -383,13 +389,37 @@ static inline uint16_t uct_ib_md_atomic_offset(uint8_t atomic_mr_id)
...
+
+    if (memh->flags == UCT_IB_MEM_FLAG_ODP) {
+        return 1;
+    }
+
+    if (flags == IBV_ACCESS_REMOTE_ATOMIC) {
+        return 0;
+    }
+
+    return flags & IBV_ACCESS_RELAXED_ORDERING;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.h:411]: this logic is very confusing and should be refactored: it should be used only when packing rkey and not during init/cleanup

File: src/uct/ib/dc/dc_mlx5.c [hunk 7]
```diff
@@ -882,11 +882,22 @@ static inline ucs_status_t uct_dc_mlx5_iface_flush_dcis(uct_dc_mlx5_iface_t *ifa
...
+    uct_ib_mlx5_md_t *md       = ucs_derived_of(iface->super.super.super.super.md,
+                                                uct_ib_mlx5_md_t);
+
     ucs_status_t status;
 
     if (comp != NULL) {
         return UCS_ERR_UNSUPPORTED;
     }
+
+    if (md->super.relaxed_order) {
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:894]: can we take this from rkey somehow?
it also could be in another PR

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 8]
```diff
@@ -942,6 +962,15 @@ static ucs_status_t uct_ib_mlx5dv_md_open(struct ibv_device *ibv_device,
...
 
     md->super.ops = &uct_ib_mlx5_md_ops;
+
+    uct_ib_md_parse_relaxed_order(&md->super, md_config);
+    if (md->super.relaxed_order) {
+        md->super.memh_struct_size = sizeof(uct_ib_mlx5_mem_t);
+    } else {
+        md->super.memh_struct_size = sizeof(uct_ib_mlx5_mem_t) -
+                                     sizeof(uct_ib_mlx5_mr_t);
+    }
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:972]: code duplication with 707-713

File: src/uct/ib/mlx5/exp/ib_exp_md.c [hunk 9]
```diff
@@ -44,7 +44,7 @@ static ucs_status_t uct_ib_mlx5_reg_key(uct_ib_md_t *md, void *address,
         return status;
     }
 
-    uct_ib_memh_init_from_mr(&memh->super, memh->mr);
+    uct_ib_memh_init_keys(md, ib_memh, access_flags,  memh->mr->lkey, memh->mr->rkey);
```
Comment (yosefe) [inline src/uct/ib/mlx5/exp/ib_exp_md.c:47]: line seems long

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 10]
```diff
@@ -665,6 +666,11 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_iface_common_t,
         goto cleanup_tm;
     }
 
+    if ((self->super.config.fence_mode == UCT_RC_FENCE_MODE_NONE) &&
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:669]: unite uct_ib_device_has_pci_atomics() and md->super.relaxed_order to
uct_ib_iface_need_fence()

**Comment by Artemy-Mellanox** [comment]
bot:pipe:retest

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 11]
```diff
@@ -42,26 +41,35 @@ typedef struct uct_ib_mlx5_dbrec_page {
...
 {
     uct_ib_mlx5_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_mlx5_mem_t);
-    ucs_status_t status;
 
-    status = uct_ib_reg_mr(md->pd, address, length, access_flags, &memh->mr);
-    if (status != UCS_OK) {
-        return status;
-    }
+    return uct_ib_verbs_reg_key_impl(md, address, length, access_flags, ib_memh,
+                                     &memh->mrs[mr_index].super, mr_index);
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:49]: how come mlx5 code is using "verbs" registration?

File: src/uct/ib/base/ib_md.c [hunk 12]
```diff
@@ -814,17 +880,23 @@ static ucs_status_t uct_ib_mkey_pack(uct_md_h uct_md, uct_mem_h uct_memh,
...
     /* create umr only if a user requested atomic access to the
      * memory region and the hardware supports it.
      */
-    if ((memh->flags & UCT_IB_MEM_ACCESS_REMOTE_ATOMIC) &&
-        !(memh->flags & UCT_IB_MEM_FLAG_ATOMIC_MR) &&
+    if ((((memh->flags & UCT_IB_MEM_ACCESS_REMOTE_ATOMIC) &&
+        !(memh->flags & UCT_IB_MEM_FLAG_ATOMIC_MR)) || md->relaxed_order) &&
         (memh != md->global_odp))
     {
+        if (memh->flags & UCT_IB_MEM_FLAG_RELAXED_ORDERING) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:892]: need to check relaxed order flag on the MR?

File: src/uct/ib/base/ib_md.c [hunk 13]
```diff
@@ -1624,6 +1702,13 @@ static ucs_status_t uct_ib_verbs_md_open(struct ibv_device *ibv_device,
         goto err_free_context;
     }
 
+    uct_ib_md_parse_relaxed_order(md, md_config);
+    if (md->relaxed_order) {
+        md->memh_struct_size = sizeof(uct_ib_verbs_mem_t);
+    } else {
+        md->memh_struct_size = sizeof(uct_ib_verbs_mem_t) - sizeof(uct_ib_mr_t);
+    }
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:1710]: how come common code looks at "verbs" struct?
this calculation may be wrong for mlx5

File: src/uct/ib/base/ib_md.c [hunk 14]
```diff
@@ -1645,11 +1730,10 @@ static ucs_status_t uct_ib_verbs_md_open(struct ibv_device *ibv_device,
 static uct_ib_md_ops_t uct_ib_verbs_md_ops = {
     .open                = uct_ib_verbs_md_open,
     .cleanup             = (uct_ib_md_cleanup_func_t)ucs_empty_function,
-    .memh_struct_size    = sizeof(uct_ib_verbs_mem_t),
     .reg_key             = uct_ib_verbs_reg_key,
     .dereg_key           = uct_ib_verbs_dereg_key,
-    .reg_atomic_key      = (uct_ib_md_reg_atomic_key_func_t)ucs_empty_function_return_unsupported,
-    .dereg_atomic_key    = (uct_ib_md_dereg_atomic_key_func_t)ucs_empty_function_return_unsupported,
+    .reg_atomic_key      = uct_ib_verbs_reg_atomic_key,
+    .dereg_atomic_key    = (uct_ib_md_dereg_atomic_key_func_t)ucs_empty_function_return_success,
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:1736]: why verbs need uct_ib_verbs_reg_atomic_key? it doesn't really do anything, since dereg_atomic_key is empty

File: src/uct/ib/base/ib_md.h [hunk 15]
```diff
@@ -93,6 +96,19 @@ typedef struct uct_ib_mem {
     uint32_t                flags;
 } uct_ib_mem_t;
 
+
+typedef union uct_ib_mr {
+    struct ibv_mr           *ib;
+} uct_ib_mr_t;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.h:102]: why needed? can't just use struct ibv_mr?

File: src/uct/ib/base/ib_md.c [hunk 16]
```diff
@@ -806,52 +809,54 @@ static ucs_status_t uct_ib_mem_dereg(uct_md_h uct_md, uct_mem_h memh)
...
 static ucs_status_t uct_ib_verbs_reg_key(uct_ib_md_t *md, void *address,
                                          size_t length, uint64_t access_flags,
-                                         uct_ib_mem_t *ib_memh)
+                                         uct_ib_mem_t *ib_memh, int mr_index)
 {
     uct_ib_verbs_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_verbs_mem_t);
 
     return uct_ib_verbs_reg_key_impl(md, address, length, access_flags,
-                                     ib_memh, &memh->mr, &memh->so_mr);
+                                     ib_memh, &memh->mrs[mr_index], mr_index);
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.c:817]: all are in use

File: src/uct/ib/base/ib_md.c [hunk 17]
```diff
@@ -806,52 +809,54 @@ static ucs_status_t uct_ib_mem_dereg(uct_md_h uct_md, uct_mem_h memh)
...
+                                                uct_ib_mem_t *ib_memh,
+                                                int mr_index)
 {
     uct_ib_verbs_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_verbs_mem_t);
 
-    if (memh->super.flags & UCT_IB_MEM_FLAG_RELAXED_ORDERING) {
-        memh->super.atomic_rkey = memh->so_mr.ib->rkey;
-        return UCS_OK;
-    } else {
+    if (mr_index != UCT_IB_MR_FENCE) {
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.c:854]: need index for mr to be base for atomic here

File: src/uct/ib/base/ib_md.c (hunk 12 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.c:892]: no

File: src/uct/ib/base/ib_md.c (hunk 14 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.c:1736]: to store strict order rkey into atomic_rkey if have

File: src/uct/ib/base/ib_md.h [hunk 18]
```diff
@@ -101,6 +101,14 @@ typedef union uct_ib_mr {
     struct ibv_mr           *ib;
 } uct_ib_mr_t;
 
+
+enum {
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.h:105]: uct_ib_mr_index_t?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 5 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:27]: this PR requires new data structure, code prior to this PR worked with old data structure,
why separate this change?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 11 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:49]: to register mtt mr mlx5 code using verbs reg, common code.

File: src/uct/ib/base/ib_md.h (hunk 15 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.h:102]: syntactic sugar, may use (struct ibv_mr **) instead

File: src/uct/ib/base/ib_md.c (hunk 13 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.c:1710]: it's verbs code

Comment (yosefe) [reply src/uct/ib/base/ib_md.c:1710]: it's src/uct/ib/**base**/ib_md.c

Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.c:1710]: right, and it hosts uct_ib_verbs_md_ops

**Comment by yosefe** [comment]
bot:pipe:retest

File: src/uct/ib/base/ib_md.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/base/ib_md.h:105]: > uct_ib_mr_index_t?

yes, or uct_ib_mr_type_t (and use mr_type instead of mr_index in all places)

File: src/uct/ib/base/ib_md.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/base/ib_md.c:817]: ok, seems problem is with function naming: it should be uct_ib_reg_key() - no "verbs" - because it's common for verbs and mlx5

File: src/uct/ib/base/ib_md.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/base/ib_md.c:854]: i see for verbs case the index is always UCT_IB_MR_ATOMIC, for which case it can be the primary?

File: src/uct/ib/base/ib_md.c (hunk 12 above)
Comment (yosefe) [reply src/uct/ib/base/ib_md.c:892]: maybe in line 889 we can check (memh->flags & UCT_IB_MEM_FLAG_RELAXED_ORDERING) instead of md->relaxed_order? instead of line 892?
since if md->relaxed_order is true, we always set UCT_IB_MEM_FLAG_RELAXED_ORDERING

File: src/uct/ib/base/ib_md.c (hunk 13 above)
Comment (yosefe) [reply src/uct/ib/base/ib_md.c:1710]: right, it's uct_ib_**verbs**_md_open
maybe define uct_ib_verbs_mem_t with zero-size array of MRs, so it will be `sizeof(uct_ib_verbs_mem_t) + sizeof(uct_ib_mr_t)`, i think it's more obvious (+add comment to explain the calculation)

File: src/uct/ib/base/ib_md.c (hunk 14 above)
Comment (yosefe) [reply src/uct/ib/base/ib_md.c:1736]: ok

File: src/uct/ib/base/ib_md.h (hunk 15 above)
Comment (yosefe) [reply src/uct/ib/base/ib_md.h:102]: let's use (struct ibv_mr **) instead

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 5 above)
Comment (yosefe) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:27]: at this point it' too late, but smaller PR's would be faster to review, it could be "preparation" PR just to change data structure

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 11 above)
Comment (yosefe) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:49]: as noted in other place - need to rename this helper function

File: src/uct/sm/scopy/knem/knem_md.c [hunk 19]
```diff
@@ -114,6 +114,7 @@ static ucs_status_t uct_knem_mem_reg_internal(uct_md_h md, void *address, size_t
 
     knem_iov[0].base = (uintptr_t) address;
     knem_iov[0].len = length;
+    //printf("%s:%d %p %zx\n", __func__, __LINE__, address, length);
```
Comment (yosefe) [inline src/uct/sm/scopy/knem/knem_md.c:117]: leftover

**Comment by yosefe** [comment]
test failure seems relevant:
http://hpc-master.lab.mtl.com:8080/blue/organizations/jenkins/ucx/detail/ucx/5492/pipeline/
```
[2020-06-12T09:22:07.334Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/mlx5/exp/ib_exp_md.c:732:28: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]

[2020-06-12T09:22:07.334Z]      .reg_key             = uct_ib_mlx5_reg_key,

[2020-06-12T09:22:07.334Z]                             ^

[2020-06-12T09:22:07.334Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/mlx5/exp/ib_exp_md.c:732:28: note: (near initialization for ‘uct_ib_mlx5_md_ops.reg_key’)

[2020-06-12T09:22:07.334Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/mlx5/exp/ib_exp_md.c:733:28: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]

[2020-06-12T09:22:07.334Z]      .dereg_key           = uct_ib_mlx5_dereg_key,

[2020-06-12T09:22:07.334Z]                             ^

[2020-06-12T09:22:07.334Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/mlx5/exp/ib_exp_md.c:733:28: note: (near initialization for ‘uct_ib_mlx5_md_ops.dereg_key’)

[2020-06-12T09:22:07.334Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/mlx5/exp/ib_exp_md.c:734:28: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]

[2020-06-12T09:22:07.334Z]      .reg_atomic_key      = uct_ib_mlx5_exp_reg_atomic_key,

[2020-06-12T09:22:07.334Z]                             ^

[2020-06-12T09:22:07.334Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/mlx5/exp/ib_exp_md.c:734:28: note: (near initialization for ‘uct_ib_mlx5_md_ops.reg_atomic_key’)

[2020-06-12T09:22:07.334Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/mlx5/exp/ib_exp_md.c:736:28: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]

[2020-06-12T09:22:07.334Z]      .reg_multithreaded   = uct_ib_mlx5_exp_reg_multithreaded,

[2020-06-12T09:22:07.334Z]                             ^

[2020-06-12T09:22:07.334Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/mlx5/exp/ib_exp_md.c:736:28: note: (near initialization for ‘uct_ib_mlx5_md_ops.reg_multithreaded’)

[2020-06-12T09:22:07.334Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/mlx5/exp/ib_exp_md.c:737:28: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]

[2020-06-12T09:22:07.334Z]      .dereg_multithreaded = uct_ib_mlx5_exp_dereg_multithreaded,

[2020-06-12T09:22:07.334Z]                             ^

[2020-06-12T09:22:07.334Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/mlx5/exp/ib_exp_md.c:737:28: note: (near initialization for ‘uct_ib_mlx5_md_ops.dereg_multithreaded’)

[2020-06-12T09:22:07.590Z]   CC       rc/base/libuct_ib_la-rc_ep.lo
```

**Comment by yosefe** [comment]
bot:pipe:retest
